### ISSUE TITLE
Integrate with the OkHttp's analytics API

### ISSUE THREAD

**Issue by bric3** [square/retrofit#2732]
What kind of issue is this?

 - [ ] Question. This issue tracker is not the place for questions. If you want to ask how to do
       something, or to understand why something isn't working the way you expect it to, use Stack
       Overflow. https://stackoverflow.com/questions/tagged/retrofit

 - [ ] Bug report. If you’ve found a bug, spend the time to write a failing test. Bugs with tests
       get fixed. Here’s an example: https://gist.github.com/swankjesse/6608b4713ad80988cdc9

 - [X] Feature Request. Start by telling us what problem you’re trying to solve. Often a solution
       already exists! Don’t send pull requests to implement new features without first getting our
       support. Sometimes we leave features out on purpose to keep the project small.

Following the analytics API that is being build within okhttp (https://github.com/square/okhttp/issues/270), I would like to capture metrics per _mapped_ path / route.

#### tl;dr

Currently it is not possible to have metrics per route in OkHttp's `EventListener` without parsing the URL against a known list of path (that has to populated from the interface or somewhere else.
This inefficient on several aspects.

I would like retrofit to communicate the route _context_ to OkHttp, inorder to access the in the EventListener.

#### Example

That is for the following interface.

```
interface FakeRest {
    @GET("/posts")
    Call<ResponseBody> posts();
    @GET("/posts/{post-id}")
    Call<ResponseBody> post(@Path("post-id") int postId);
    @GET("/posts/{post-id}/comments")
    Call<ResponseBody> postComments(@Path("post-id") int postId);
}
```

I would like to access the method or a pre-built request/call context that the `EventListener` can access. I've played around with the Retrofit builder using `Call.Factory`, `CallAdapter.Factory` but unfortunately I didn't find a way to capture this information and to pass it in the HTTP Call Request.

I noticed the `tag` field in OkHttp's `Request`, I know it's not documented for this kind of usage, but it would helped me to bootstrap the idea.

#### What I tried :

Have a built retrofit.

```java
Retrofit retrofit = new Retrofit.Builder().baseUrl("https://jsonplaceholder.typicode.com/").build();
```

_Decorate_ each `CallAdapter.Factory` in a new Retrofit created form the original one, and capture path information from annotations (possibly use `baseUrl` as well).

```
Retrofit.Builder newBuilder = retrofit.newBuilder();
for (Factory callAdapterFactory : retrofit.callAdapterFactories()) {
    newBuilder.addCallAdapterFactory(new Factory() {
        @Override
        public CallAdapter<Object, Object> get(Type returnType, Annotation[] annotations, Retrofit retrofit) {
            String normalizedPath = normalizedPathFrom(annotations);
            CallAdapter<Object, Object> callAdapter = (CallAdapter<Object, Object>) callAdapterFactory.get(returnType, annotations, retrofit);
            return new CallAdapter<Object, Object>() {
                ...
                @Override
                public Object adapt(Call<Object> call) {
                    return callAdapter.adapt(wrapCall(call));
                }
            };
        }
    });
}
retrofit = newBuilder.build();
```

But the thing here is that I don't get how to pass it to an OkHttp `Request`. The `Call` received in the `adapt` method cannot be modified, wrapping it won't work, because internally (in `OkHttpCall`) retrofit will build his own immutable request that will be used for executing or enqueuing the actual request. Cloning is of course out of question, and duplicating it appears impossible because the constructors are not public.

------------

Using the `Call.Factory` is no help, because the this interface receives a pre-built request, so I cannot introduce a `Call.Factory` decorator that capture the necessary annotations and pass them in the request tag.

To be able to customize request, I believe that either `RequestBuilder` could be made an interface (and making the according modifications), or that `RequestBuilder` accept some sort of `RequestCustomizer` that can access the **route context**, either headers, path, query param names, interface method name, etc.

**Comment by adriancole** [comment]
@jkschneider have you run across any approaches here? Only thing that came to mind to me are..

* request tag wrapper
  * coordinated where retrofit wraps the any existing request.tag with something including the route, the okhttp interceptor unwraps that after metrics start
* synthetic http header
  * similar but using a request header instead

**Comment by bric3** [comment]
@adriancole On thing about the synthetic header, is that I would like to avoid okhttp to send it, as far as I know there's no such notion in OkHttp as of today.

**Comment by jkschneider** [comment]
@adriancole For some reason that I can't remember we avoided the analytics API altogether (perhaps it was a backwards compatibility concern?). Doesn't appear that Micrometer's [current implementation](https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/okhttp3/OkHttpMetricsEventListenerTest.java) is aware of the parameterized URI either.

**Comment by bric3** [comment]
@adriancole I took the time to play a bit with the call adapter factory and ok event listener. I have a solution that can capture annotations and store them in the request _tag_, these annotations can be read in the `EventListener`.

=> https://github.com/bric3/retrofit-okhttp-accounting/

What I don't like with this solution is that, this a quite hacky one

1. The Call wrapper is relying on the fact that calling `.request()` on the actual `call` before executing or enqueuing the call will return the request that will be used in this operations, it could be a copy for example, since the object is _almost immutable this is probably unlikely but given that _tag_ can be a mutable object I don't know if that is a lasting approach. 

2. The `tag` type is `Object` and has _self_ (the request itself) value by default; so the Call wrapper has to cast the `tag` to the needed type, and the value has to be set at the request creation time because a `Request` is immutable. Currently this is done in the `CallFactory`, where the request is modified with a builder then passed to ok http client. Design wise that's wrong, that's splitting concerns in different places. What if the `tag` has to be used for something else.

3. OkHttp `Request` builder documents the `tag` as way to cancel the request. I'm not sure how this is used, but using `tag` to achieve accounting or monitoring feels like a misappropriate use of the `tag`.

   > Attaches {@code tag} to the request. It can be used later to cancel the request. If the tag is unspecified or null, the request is canceled by using the request itself as the tag.

I think this would be very welcome improvement in both retrofit and okhttp to have this kind of feature.
I would like a simpler approach without having to rely on a `CallAdapterFactory` and a custom `CalFactory`

1. OkHttp' request could benefit from a field dedicated to contextual information.
2. Retrofit should have a request factory, or some kind of _request customizer_, or _request context customizer_, that could be leveraged with interface methods annotations like `@DELETE`.

**Comment by nicktrav** [comment]
Just adding in my two cents here. We'd also love to be able to get at the method that was called on the interface.

Talking to @swankjesse about this internally, it sounds like he has some ideas for this that might involve being able to add a `retrofit.Call` into the `okhttp.Response` tags, and then being able to pluck that metadata back out for use later.

**Comment by swankjesse** [comment]
https://github.com/square/retrofit/pull/2899
